File: backend/services/stigmer-server/pkg/domain/workflowexecution/controller/update_status.go
package workflowexecution

import (
	"context"

	"github.com/rs/zerolog/log"
	workflowexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflowexecution/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	grpclib "github.com/stigmer/stigmer/backend/libs/go/grpc"
	"github.com/stigmer/stigmer/backend/libs/go/grpc/request/pipeline"
	"github.com/stigmer/stigmer/backend/libs/go/store"
	"google.golang.org/protobuf/proto"
)

// UpdateStatus updates execution status during workflow execution
//
// Used by workflow-runner to send progressive status updates (tasks, phase, etc.)
// This RPC is optimized for frequent status updates and merges status fields with existing state.
//
// Pipeline Steps:
// 1. ValidateInput - Validate execution_id and status are provided
// 2. LoadExisting - Load existing execution from DB
// 3. BuildNewStateWithStatus - Merge status updates from input
// 4. Persist - Save to database
// 5. BroadcastToStreams - Push update to active Go channels (ADR 011)
//
// Note: Compared to Stigmer Cloud, OSS excludes:
// - Authorize step (no multi-tenant auth in OSS)
// - PublishToRedis step (no Redis in OSS - uses in-memory Go channels instead per ADR 011)
// - Publish step (no event publishing in OSS)
func (c *WorkflowExecutionController) UpdateStatus(ctx context.Context, input *workflowexecutionv1.WorkflowExecutionUpdateStatusInput) (*workflowexecutionv1.WorkflowExecution, error) {
	// Create request context with input
	reqCtx := pipeline.NewRequestContext(ctx, input)

	// Build pipeline
	p := pipeline.NewPipeline[*workflowexecutionv1.WorkflowExecutionUpdateStatusInput]("workflowexecution-update-status").
		AddStep(newValidateUpdateStatusInputStep()).
		AddStep(newLoadExistingExecutionStep(c.store)).
		AddStep(newBuildNewStateWithStatusStep()).
		AddStep(newPersistExecutionStep(c.store)).
		AddStep(newPersistEventsStep(c.store)).
		AddStep(newBroadcastToStreamsStep(c.streamBroker)).
		Build()

	// Execute pipeline
	if err := p.Execute(reqCtx); err != nil {
		return nil, err
	}

	// Return updated execution from context
	execution, ok := reqCtx.Get("execution").(*workflowexecutionv1.WorkflowExecution)
	if !ok {
		return nil, grpclib.InternalError(nil, "execution not found in context after pipeline")
	}

	return execution, nil
}

// mergePendingByChild performs a per-child upsert of a workflow status list
// (pending_approvals or pending_file_reviews): it drops the existing entries that
// belong to scopeChildID and appends the incoming entries (which must all belong
// to scopeChildID). Every other child's entries are preserved untouched, so
// parallel child agents surfacing/clearing their own gates never clobber each
// other. An empty incoming slice therefore clears just scopeChildID's entries.
//
// childOf extracts an entry's owning child_agent_execution_id. Kept generic so the
// approval and file-review lists share exactly one merge semantic.
func mergePendingByChild[T any](existing, incoming []T, childOf func(T) string, scopeChildID string) []T {
	merged := make([]T, 0, len(existing)+len(incoming))
	for _, e := range existing {
		if childOf(e) != scopeChildID {
			merged = append(merged, e)
		}
	}
	return append(merged, incoming...)
}

// ValidateUpdateStatusInputStep validates the input for UpdateStatus
type ValidateUpdateStatusInputStep struct{}

func newValidateUpdateStatusInputStep() *ValidateUpdateStatusInputStep {
	return &ValidateUpdateStatusInputStep{}
}

func (s *ValidateUpdateStatusInputStep) Name() string {
	return "ValidateUpdateStatusInput"
}

func (s *ValidateUpdateStatusInputStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.WorkflowExecutionUpdateStatusInput]) error {
	input := ctx.Input()

	if input == nil {
		return grpclib.InvalidArgumentError("input is required")
	}

	if input.ExecutionId == "" {
		return grpclib.InvalidArgumentError("execution_id is required")
	}

	if input.Status == nil {
		return grpclib.InvalidArgumentError("status is required")
	}

	log.Debug().
		Str("execution_id", input.ExecutionId).
		Msg("Validated UpdateStatus input")

	return nil
}

// LoadExistingExecutionStep loads the existing execution from database
type LoadExistingExecutionStep struct {
	store store.Store
}

func newLoadExistingExecutionStep(store store.Store) *LoadExistingExecutionStep {
	return &LoadExistingExecutionStep{store: store}
}

func (s *LoadExistingExecutionStep) Name() string {
	return "LoadExistingExecution"
}

func (s *LoadExistingExecutionStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.WorkflowExecutionUpdateStatusInput]) error {
	input := ctx.Input()
	executionID := input.ExecutionId

	log.Debug().
		Str("execution_id", executionID).
		Msg("Loading existing execution")

	existing := &workflowexecutionv1.WorkflowExecution{}
	if err := s.store.GetResource(ctx.Context(), apiresourcekind.ApiResourceKind_workflow_execution, executionID, existing); err != nil {
		return grpclib.NotFoundError("WorkflowExecution", executionID)
	}

	// Store existing execution in context for merge step
	ctx.Set("existingExecution", existing)

	log.Debug().
		Str("execution_id", executionID).
		Str("phase", existing.Status.GetPhase().String()).
		Msg("Loaded existing execution")

	return nil
}

// BuildNewStateWithStatusStep merges status updates from input with existing execution
//
// This step follows the Java implementation's merge logic:
// - Replaces tasks array
// - Updates phase, output, error, timestamps if provided
// - Preserves spec from existing execution (does NOT update spec)
type BuildNewStateWithStatusStep struct{}

func newBuildNewStateWithStatusStep() *BuildNewStateWithStatusStep {
	return &BuildNewStateWithStatusStep{}
}

func (s *BuildNewStateWithStatusStep) Name() string {
	return "BuildNewStateWithStatus"
}

func (s *BuildNewStateWithStatusStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.WorkflowExecutionUpdateStatusInput]) error {
	input := ctx.Input()
	existing, ok := ctx.Get("existingExecution").(*workflowexecutionv1.WorkflowExecution)
	if !ok {
		return grpclib.InternalError(nil, "existing execution not found in context")
	}

	// Start with existing execution as base (cloning)
	updated := proto.Clone(existing).(*workflowexecutionv1.WorkflowExecution)

	// Ensure status is initialized
	if updated.Status == nil {
		updated.Status = &workflowexecutionv1.WorkflowExecutionStatus{}
	}

	requestStatus := input.Status

	// CRITICAL: Merge status from input (for progressive updates from workflow-runner)
	// Following Java implementation's merge strategy

	// Merge tasks (replace with latest from request)
	if len(requestStatus.Tasks) > 0 {
		updated.Status.Tasks = requestStatus.Tasks
	}

	// Update phase (if provided)
	if requestStatus.Phase != workflowexecutionv1.ExecutionPhase_EXECUTION_PHASE_UNSPECIFIED {
		updated.Status.Phase = requestStatus.Phase
	}

	// Update output (if provided)
	if requestStatus.Output != nil {
		updated.Status.Output = requestStatus.Output
	}

	// Update error (if provided)
	if requestStatus.Error != "" {
		updated.Status.Error = requestStatus.Error
	}

	// Update timestamps (if provided)
	if requestStatus.StartedAt != "" {
		updated.Status.StartedAt = requestStatus.StartedAt
	}
	if requestStatus.CompletedAt != "" {
		updated.Status.CompletedAt = requestStatus.CompletedAt
	}

	// Update temporal workflow ID (if provided)
	if requestStatus.TemporalWorkflowId != "" {
		updated.Status.TemporalWorkflowId = requestStatus.TemporalWorkflowId
	}

	// Update cost/token totals (if provided, runner accumulates across tasks)
	if requestStatus.TotalCostMicros > 0 {
		updated.Status.TotalCostMicros = requestStatus.TotalCostMicros
	}
	if requestStatus.TotalInputTokens > 0 {
		updated.Status.TotalInputTokens = requestStatus.TotalInputTokens
	}
	if requestStatus.TotalOutputTokens > 0 {
		updated.Status.TotalOutputTokens = requestStatus.TotalOutputTokens
	}

	// Guarded, per-child merge: only touch pending_approvals / pending_file_reviews
	// when explicitly requested, and even then replace only the entries for the
	// scoped child (pending_update_child_agent_execution_id), preserving every
	// sibling child's entries. This both prevents event emissions (which don't
	// include these lists) from clobbering active gates set by call-agent-status,
	// and prevents parallel child agents from clobbering each other's gates. A
	// scoped write with an empty incoming list clears just that child's entries.
	scopeChildID := input.GetPendingUpdateChildAgentExecutionId()
	if input.UpdatePendingApprovals {
		updated.Status.PendingApprovals = mergePendingByChild(
			updated.Status.GetPendingApprovals(),
			requestStatus.GetPendingApprovals(),
			func(pa *workflowexecutionv1.WorkflowPendingApproval) string { return pa.GetChildAgentExecutionId() },
			scopeChildID,
		)
	}
	if input.UpdatePendingFileReviews {
		updated.Status.PendingFileReviews = mergePendingByChild(
			updated.Status.GetPendingFileReviews(),
			requestStatus.GetPendingFileReviews(),
			func(fr *workflowexecutionv1.WorkflowPendingFileReview) string { return fr.GetChildAgentExecutionId() },
			scopeChildID,
		)
	}

	log.Debug().
		Str("execution_id", input.ExecutionId).
		Str("phase", updated.Status.Phase.String()).
		Int("tasks_count", len(updated.Status.Tasks)).
		Int("pending_approvals_count", len(updated.Status.PendingApprovals)).
		Int("pending_file_reviews_count", len(updated.Status.PendingFileReviews)).
		Msg("Merged status fields")

	// Store merged execution in context for persist step
	ctx.Set("execution", updated)

	return nil
}

// PersistExecutionStep saves the execution to database
type PersistExecutionStep struct {
	store store.Store
}

func newPersistExecutionStep(store store.Store) *PersistExecutionStep {
	return &PersistExecutionStep{store: store}
}

func (s *PersistExecutionStep) Name() string {
	return "PersistExecution"
}

func (s *PersistExecutionStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.WorkflowExecutionUpdateStatusInput]) error {
	execution, ok := ctx.Get("execution").(*workflowexecutionv1.WorkflowExecution)
	if !ok {
		return grpclib.InternalError(nil, "execution not found in context")
	}

	executionID := execution.Metadata.Id

	if err := s.store.SaveResource(ctx.Context(), apiresourcekind.ApiResourceKind_workflow_execution, executionID, execution); err != nil {
		log.Error().
			Err(err).
			Str("execution_id", executionID).
			Msg("Failed to persist execution with updated status")
		return grpclib.InternalError(err, "failed to update execution status")
	}

	log.Info().
		Str("execution_id", executionID).
		Str("phase", execution.Status.Phase.String()).
		Msg("Successfully updated execution status")

	return nil
}

// PersistEventsStep appends workflow execution events to the event log.
// Events are supplementary — a failure here logs a warning but does NOT
// fail the pipeline because status persistence already succeeded.
type PersistEventsStep struct {
	store store.Store
}

func newPersistEventsStep(store store.Store) *PersistEventsStep {
	return &PersistEventsStep{store: store}
}

func (s *PersistEventsStep) Name() string {
	return "PersistEvents"
}

func (s *PersistEventsStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.WorkflowExecutionUpdateStatusInput]) error {
	input := ctx.Input()
	events := input.GetEvents()
	if len(events) == 0 {
		return nil
	}

	executionID := input.ExecutionId

	records := make([]*store.WorkflowExecutionEventRecord, 0, len(events))
	for _, evt := range events {
		data, err := proto.Marshal(evt)
		if err != nil {
			log.Warn().
				Err(err).
				Str("execution_id", executionID).
				Uint64("sequence_number", evt.GetSequenceNumber()).
				Msg("Failed to marshal event, skipping batch")
			return nil
		}

		records = append(records, &store.WorkflowExecutionEventRecord{
			ExecutionID:    executionID,
			SequenceNumber: int64(evt.GetSequenceNumber()),
			EventType:      evt.GetEventType().String(),
			TaskName:       evt.GetTaskName(),
			Data:           data,
		})
	}

	appended, err := s.store.AppendWorkflowExecutionEvents(ctx.Context(), executionID, records)
	if err != nil {
		log.Warn().
			Err(err).
			Str("execution_id", executionID).
			Int("event_count", len(records)).
			Msg("Failed to persist execution events (non-fatal)")
		return nil
	}

	log.Debug().
		Str("execution_id", executionID).
		Int("appended", appended).
		Msg("Persisted execution events")

	return nil
}

// BroadcastToStreamsStep broadcasts the execution update to all active subscribers
//
// This implements the "Daemon (Streaming): Pushes message to active Go Channels" step
// from ADR 011 Write Path.
//
// After persisting to SQLite, the daemon must push updates to in-memory channels
// so that Subscribe() streams can receive updates in real-time without polling.
type BroadcastToStreamsStep struct {
	broker *StreamBroker
}

func newBroadcastToStreamsStep(broker *StreamBroker) *BroadcastToStreamsStep {
	return &BroadcastToStreamsStep{broker: broker}
}

func (s *BroadcastToStreamsStep) Name() string {
	return "BroadcastToStreams"
}

func (s *BroadcastToStreamsStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.WorkflowExecutionUpdateStatusInput]) error {
	execution, ok := ctx.Get("execution").(*workflowexecutionv1.WorkflowExecution)
	if !ok {
		return grpclib.InternalError(nil, "execution not found in context")
	}

	// Broadcast to all active subscribers
	s.broker.Broadcast(execution)

	log.Debug().
		Str("execution_id", execution.Metadata.Id).
		Int("subscribers", s.broker.GetSubscriberCount(execution.Metadata.Id)).
		Msg("Broadcasted execution update to subscribers")

	return nil
}

File: backend/services/stigmer-server/src/domain/workflowexecution/update-status.ts
/**
 * UpdateStatus — ports controller/update_status.go: the runner's
 * progressive status-update RPC (merge + persist + event append + stream
 * broadcast — the OSS replacement for cloud's Redis write path).
 *
 * Chain per Go: ValidateUpdateStatusInput → merge+persist → PersistEvents
 * → BroadcastToStreams. Go still does the merge as a separate
 * LoadExisting step followed by a plain SaveResource — a lost-update
 * window under concurrent updates. This port runs the SAME merge body
 * inside the store's atomic `updateResource` instead (sub-project DD-001,
 * owner-ratified): Go's own agentexecution domain already calls that
 * discipline load-bearing, and this domain's per-child pending-gate merge
 * is exactly the concurrent-children write that the window can corrupt.
 * Wire-identical in sequential flows; strictly safer under concurrency;
 * the Go-side race is filed as an OSS issue.
 *
 * applyUpdateStatusMerge is the merge body run inside the updateResource
 * closure (and exercised directly by the guard tests), mirroring the
 * cloud handler's strategy: presence-guarded field merges (tasks replaced
 * wholesale, counters only when > 0), the flag-gated per-child
 * pending_approvals / pending_file_reviews merge, and the
 * phase-transition-only statusAudit bump (heartbeats must not perpetually
 * re-sort long-running executions above new ones in the recents sidebar).
 *
 * Versus Stigmer Cloud, OSS excludes the Authorize, PublishToRedis, and
 * Publish steps (broadcast rides in-memory channels per ADR 011).
 */
import { create, toBinary } from "@bufbuild/protobuf";
import { timestampNow } from "@bufbuild/protobuf/wkt";

import type { WorkflowExecution } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import {
  WorkflowExecutionSchema,
  WorkflowExecutionStatusSchema,
} from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import { WorkflowExecutionCommandController } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/command_pb";
import { ExecutionPhase } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/enum_pb";
import { WorkflowEventType } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/event_pb";
import { WorkflowExecutionEventSchema } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/event_pb";
import type { WorkflowExecutionUpdateStatusInput } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/io_pb";
import {
  ApiResourceAuditInfoSchema,
  ApiResourceAuditSchema,
} from "@stigmer/protos/ai/stigmer/commons/apiresource/status_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import type { Logger } from "../../boot/logger.js";
import {
  internalError,
  invalidArgumentError,
  notFoundError,
} from "../../pipeline/errors.js";
import { newPipeline } from "../../pipeline/pipeline.js";
import type { Authorizer } from "../../extensions/authorizer.js";
import { newAuthorizeStep } from "../../pipeline/steps/authorize.js";
import type { CallerIdentity } from "../../extensions/identity.js";
import { RequestContext } from "../../pipeline/request-context.js";
import { ResourceNotFoundError } from "../../store/interface.js";
import type {
  Store,
  WorkflowExecutionEventRecord,
} from "../../store/interface.js";

import type { StreamBroker } from "./stream-broker.js";

export interface UpdateStatusDeps {
  readonly store: Store;
  readonly logger: Logger;
  /** The composed authorization seam — the Authorize step at position 1 of every chain calls it (O2, DD-007 §3). */
  readonly authorizer: Authorizer;
  readonly broker: StreamBroker;
}

type UpdateStatusDesc =
  typeof WorkflowExecutionCommandController.method.updateStatus.input;

const EXECUTION_KEY = "execution";

export async function updateStatus(
  deps: UpdateStatusDeps,
  input: WorkflowExecutionUpdateStatusInput,
  identity: CallerIdentity,
): Promise<WorkflowExecution> {
  const reqCtx = new RequestContext(
    WorkflowExecutionCommandController.method.updateStatus.input,
    input,
    identity,
    ApiResourceKind.workflow_execution,
  );
  await newPipeline<UpdateStatusDesc>(
    "workflowexecution-update-status",
    deps.logger,
  )
    .addStep(
      newAuthorizeStep(
        WorkflowExecutionCommandController.method.updateStatus,
        deps.authorizer,
      ),
    )
    .addStep({
      name: "ValidateUpdateStatusInput",
      execute(ctx) {
        if (ctx.input.executionId === "") {
          throw invalidArgumentError("execution_id is required");
        }
        if (ctx.input.status === undefined) {
          throw invalidArgumentError("status is required");
        }
      },
    })
    .addStep({
      name: "MergeAndPersistExecution",
      async execute(ctx) {
        let updated: WorkflowExecution;
        try {
          updated = await deps.store.updateResource(
            ApiResourceKind.workflow_execution,
            ctx.input.executionId,
            WorkflowExecutionSchema,
            (execution) => {
              applyUpdateStatusMerge(execution, ctx.input);
            },
          );
        } catch (error) {
          if (error instanceof ResourceNotFoundError) {
            // Go's LoadExistingExecution answers this exact NotFound.
            throw notFoundError("WorkflowExecution", ctx.input.executionId);
          }
          throw internalError(error, "failed to update execution status");
        }
        ctx.set(EXECUTION_KEY, updated);
        deps.logger.info("Successfully updated execution status", {
          executionId: ctx.input.executionId,
          phase:
            ExecutionPhase[
              updated.status?.phase ??
                ExecutionPhase.EXECUTION_PHASE_UNSPECIFIED
            ],
        });
      },
    })
    .addStep({
      name: "PersistEvents",
      async execute(ctx) {
        await persistEvents(deps, ctx.input);
      },
    })
    .addStep({
      name: "BroadcastToStreams",
      execute(ctx) {
        const execution = ctx.get(EXECUTION_KEY);
        if (execution === undefined) {
          throw internalError(
            new Error("execution not found in context"),
            "execution not found in context",
          );
        }
        // Push to active subscribers AFTER the persist commits (ADR 011
        // write path) — the ordering subscribe's register-before-snapshot
        // guarantee builds on.
        deps.broker.broadcast(execution as WorkflowExecution);
      },
    })
    .build()
    .execute(reqCtx);

  const merged = reqCtx.get(EXECUTION_KEY);
  if (merged === undefined) {
    throw internalError(
      new Error("execution not found in context after pipeline"),
      "execution not found in context after pipeline",
    );
  }
  return merged as WorkflowExecution;
}

/**
 * mergePendingByChild — a per-child upsert of a workflow status list
 * (pending_approvals or pending_file_reviews): drops the existing entries
 * that belong to scopeChildId and appends the incoming entries (which
 * must all belong to scopeChildId). Every other child's entries are
 * preserved untouched, so parallel child agents surfacing/clearing their
 * own gates never clobber each other. An empty incoming list therefore
 * clears just scopeChildId's entries.
 *
 * childOf extracts an entry's owning child_agent_execution_id. Kept
 * generic so the approval and file-review lists share exactly one merge
 * semantic (Go's mergePendingByChild).
 */
export function mergePendingByChild<T>(
  existing: readonly T[],
  incoming: readonly T[],
  childOf: (entry: T) => string,
  scopeChildId: string,
): T[] {
  const merged: T[] = [];
  for (const entry of existing) {
    if (childOf(entry) !== scopeChildId) {
      merged.push(entry);
    }
  }
  merged.push(...incoming);
  return merged;
}

/**
 * The merge body run inside the updateResource closure — Go
 * BuildNewStateWithStatusStep field-for-field. Mutates `execution` in
 * place over the freshly-loaded, write-locked state. Synchronous by
 * contract (the store's modify callback), which is also what makes it
 * directly unit-testable.
 */
export function applyUpdateStatusMerge(
  execution: WorkflowExecution,
  input: WorkflowExecutionUpdateStatusInput,
): void {
  if (execution.status === undefined) {
    execution.status = create(WorkflowExecutionStatusSchema);
  }
  const status = execution.status;
  const requestStatus = input.status;
  if (requestStatus === undefined) {
    return;
  }

  // The phase-transition check compares against the phase BEFORE this
  // merge writes it (Go compares to `existing`, its pre-clone snapshot).
  const previousPhase = status.phase;

  // Tasks: replaced wholesale with the runner's latest complete set.
  if (requestStatus.tasks.length > 0) {
    status.tasks = requestStatus.tasks;
  }

  if (requestStatus.phase !== ExecutionPhase.EXECUTION_PHASE_UNSPECIFIED) {
    status.phase = requestStatus.phase;
  }
  if (requestStatus.output !== undefined) {
    status.output = requestStatus.output;
  }
  if (requestStatus.error !== "") {
    status.error = requestStatus.error;
  }
  if (requestStatus.startedAt !== "") {
    status.startedAt = requestStatus.startedAt;
  }
  if (requestStatus.completedAt !== "") {
    status.completedAt = requestStatus.completedAt;
  }
  if (requestStatus.temporalWorkflowId !== "") {
    status.temporalWorkflowId = requestStatus.temporalWorkflowId;
  }

  // Cost/token totals: the runner sends accumulated totals; zero means
  // "no update", never "reset".
  if (requestStatus.totalCostMicros > 0n) {
    status.totalCostMicros = requestStatus.totalCostMicros;
  }
  if (requestStatus.totalInputTokens > 0n) {
    status.totalInputTokens = requestStatus.totalInputTokens;
  }
  if (requestStatus.totalOutputTokens > 0n) {
    status.totalOutputTokens = requestStatus.totalOutputTokens;
  }

  // Guarded, per-child merge: only touch pending_approvals /
  // pending_file_reviews when explicitly requested, and even then replace
  // only the scoped child's entries. This both prevents event emissions
  // (which don't include these lists) from clobbering active gates set by
  // call-agent-status, and prevents parallel child agents from clobbering
  // each other's gates.
  const scopeChildId = input.pendingUpdateChildAgentExecutionId;
  if (input.updatePendingApprovals) {
    status.pendingApprovals = mergePendingByChild(
      status.pendingApprovals,
      requestStatus.pendingApprovals,
      (entry) => entry.childAgentExecutionId,
      scopeChildId,
    );
  }
  if (input.updatePendingFileReviews) {
    status.pendingFileReviews = mergePendingByChild(
      status.pendingFileReviews,
      requestStatus.pendingFileReviews,
      (entry) => entry.childAgentExecutionId,
      scopeChildId,
    );
  }

  // Only bump statusAudit.updatedAt on phase transitions — not on every
  // task-progress heartbeat from the runner. This prevents long-running
  // executions from perpetually sorting above freshly created ones in the
  // recents sidebar (ActivityQueryController orders by this field).
  // Mirrors the cloud's WorkflowExecutionUpdateStatusHandler exactly.
  const isPhaseTransition =
    requestStatus.phase !== ExecutionPhase.EXECUTION_PHASE_UNSPECIFIED &&
    requestStatus.phase !== previousPhase;
  if (isPhaseTransition) {
    const audit = status.audit ?? create(ApiResourceAuditSchema);
    status.audit = audit;
    const statusAudit = audit.statusAudit ?? create(ApiResourceAuditInfoSchema);
    audit.statusAudit = statusAudit;
    statusAudit.updatedAt = timestampNow();
    statusAudit.event = "updated";
  }
}

/**
 * PersistEvents — appends the input's events to the event log. Events are
 * supplementary: a failure here does NOT fail the pipeline because status
 * persistence already succeeded. A real append failure means timeline
 * data loss, so it logs at error level; skipped duplicates are the
 * expected result of the runner's idempotent batch retries and only log
 * at debug (the store's INSERT OR IGNORE first-writer-wins, oss#308).
 */
async function persistEvents(
  deps: UpdateStatusDeps,
  input: WorkflowExecutionUpdateStatusInput,
): Promise<void> {
  if (input.events.length === 0) {
    return;
  }
  const executionId = input.executionId;

  const records: WorkflowExecutionEventRecord[] = [];
  for (const event of input.events) {
    let data: Uint8Array;
    try {
      data = toBinary(WorkflowExecutionEventSchema, event);
    } catch (error) {
      deps.logger.error(
        "Failed to marshal event — dropping batch (non-fatal, timeline data lost)",
        {
          executionId,
          sequenceNumber: event.sequenceNumber,
          eventCount: input.events.length,
          error: error instanceof Error ? error.message : String(error),
        },
      );
      return;
    }
    records.push({
      executionId,
      sequenceNumber: Number(event.sequenceNumber),
      eventType: WorkflowEventType[event.eventType] ?? "",
      taskName: event.taskName,
      data,
      createdAt: "",
    });
  }

  let appended: number;
  try {
    appended = await deps.store.appendWorkflowExecutionEvents(
      executionId,
      records,
    );
  } catch (error) {
    deps.logger.error(
      "Failed to persist execution events (non-fatal, timeline data lost)",
      {
        executionId,
        submitted: records.length,
        error: error instanceof Error ? error.message : String(error),
      },
    );
    return;
  }

  const skipped = records.length - appended;
  if (skipped > 0) {
    deps.logger.debug(
      "Persisted execution events (idempotent retry skipped already-persisted sequences)",
      { executionId, appended, skippedDuplicates: skipped },
    );
  } else {
    deps.logger.debug("Persisted execution events", { executionId, appended });
  }
}
